sources/api/early-boot-config/src/provider/vmware.rs

```diff
@@ -125,23 +134,146 @@ impl VmwareDataProvider {
+                let mut decoder = base64::read::DecoderReader::new(&mut reader, base64::STANDARD);
+
+                let mut decoded_bytes = Vec::new();
+                decoder
+                    .read_to_end(&mut decoded_bytes)
+                    .context(error::StreamingBase64Decode {
+                        what: "guestinfo user data",
+                    })?;
+
+                // Decompresses the data if it is gzip'ed
+                expand_slice_maybe(&decoded_bytes).context(error::Decompression {
+                    what: "guestinfo user data",
+                })?
+            }
+
+            UserDataEncoding::Raw => {
+                String::from_utf8(user_data_bytes).context(error::InvalidUtf8 {
+                    what: Self::GUESTINFO_USERDATA,
+                })?
+            }
+        };
+
+        let json = SettingsJson::from_toml_str(&user_data_string, "user data")?;
+        Ok(Some(json))
+    }
+
+    /// Request a key's value from guestinfo
+    // TODO: Figure out if it's possible to re-use the backdoor or ERPC channel (mutable
+    // references everywhere...)
+    fn backdoor_get_bytes(key: &str) -> Result<Option<Vec<u8>>> {
+        // Probe and access the VMWare backdoor.  `kernel lockdown(7)` may block "privileged"
+        // mode because of its use of `iopl()`.  According the the bug report below, in theory
+        // "privileged" mode is more reliable than "unprivileged" but both provide access to
+        // the same data so we fall back to "unprivileged" if the first call fails.
+        // https://github.com/lucab/vmw_backdoor-rs/issues/6
+        let mut backdoor = vmw_backdoor::probe_backdoor_privileged()
+            .or_else(|e| {
+                warn!(
+                    "Unable to access backdoor via privileged mode, using unprivileged: {}",
+                    e
+                );
+                vmw_backdoor::probe_backdoor()
+            })
+            .context(error::Backdoor {
+                op: "probe and acquire access",
+            })?;
+
+        let mut erpc = backdoor.open_enhanced_chan().context(error::Backdoor {
+            op: "open eRPC channel",
+        })?;
+
+        erpc.get_guestinfo(key.as_bytes())
+            .context(error::GuestInfo { what: key })
+    }
 }
 
 impl PlatformDataProvider for VmwareDataProvider {
     fn platform_data(&self) -> std::result::Result<Vec<SettingsJson>, Box<dyn std::error::Error>> {
         let mut output = Vec::new();
 
-        match Self::user_data() {
-            Err(e) => return Err(e).map_err(Into::into),
-            Ok(None) => warn!("No user data found."),
+        match Self::guestinfo_user_data() {
+            Err(e) => {
+                error!("Error retrieving user data from guestinfo: {}", e);
+            }
+            Ok(None) => warn!("No user data found via guestinfo"),
             Ok(Some(s)) => output.push(s),
         }
+
+        // If we didn't find any user data in guestinfo, try CD-ROM
```

Comment:
I'm curious what others think here.  If I specified user data in two locations, I'd expect both of them to be obeyed (in some order, but that's a different problem) rather than one taking precedence and ignoring the other.  Perhaps I have a common user data CD that I use in all launches, and then I use guestinfo for more specific settings in some launches, or whatever.

Reply:
This seems good now, but I'm wondering where you think we should document this?

Reply:
Creating an issue for that now.  We can probably comment on it here, but I the "main" documentation should happen in the variant's top level `README` unless you think it belongs somewhere else.